### PR TITLE
test: :white_check_mark: add tests crate `db`

### PR DESCRIPTION
**Pull request by jbcaron**
## Pull Request type

- Testing

## What is the current behavior?

Resolves: #NA

## What is the new behavior?

- Trait Implementations:
  - Implemented the Default trait on primitives to allow for the creation of default instances during testing.
  - Implemented Eq and PartialEq traits on primitives to enable equality checks within tests.
- Unit Tests:
  - Block Header Storage and Retrieval:
    - Added tests to ensure that block headers are correctly stored and can be accurately retrieved.
  - Transaction Storage and Retrieval:
    - Added tests to validate the storage of transactions and their subsequent retrieval, ensuring that the data remains consistent.
  - Receipts Storage and Retrieval:
    - Introduced tests for storing and retrieving receipts, confirming that all relevant details are preserved.
  - Transaction Search:
    - Implemented tests to verify that the transaction search functionality works as expected, returning the correct transaction data based on provided search criteria.
- Pending Block Management:
  - Added tests to cover the management of pending blocks, ensuring that pending block is handled properly in various scenarios, such as during storage and retrieval processes.

## Does this introduce a breaking change?

No

### CODE REVIEW

File: crates/client/db/tests/test_block.rs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    DeployAccountTransactionV3, DeployTransaction, InvokeTransactionV0, InvokeTransactionV1, InvokeTransactionV3,
+    L1HandlerTransaction,
+};
+use starknet_types_core::felt::Felt;
+
+#[tokio::test]
+async fn test_chain_info() {
+    let db = temp_db().await;
+    let chain_config = db.backend().chain_config();
+    assert_eq!(chain_config.chain_id, ChainConfig::test_config().chain_id);
```
Comment (antiyro) [inline crates/client/db/tests/test_block.rs:26]: Could we add some context on those asserts?

File: crates/client/db/tests/test_block.rs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+    let block_pending = block_pending_one();
+    backend.store_block(block_pending.clone(), state_diff_pending_one(), vec![]).unwrap();
+
+    let tx_hash_1 = block_pending.info.tx_hashes()[1];
+    assert_eq!(backend.find_tx_hash_block_info(&tx_hash_1).unwrap().unwrap(), (block_pending.info.clone(), TxIndex(1)));
+    assert_eq!(backend.find_tx_hash_block(&tx_hash_1).unwrap().unwrap(), (block_pending, TxIndex(1)));
+}
+
+fn block_zero() -> DeoxysMaybePendingBlock {
```
Comment (antiyro) [inline crates/client/db/tests/test_block.rs]: maybe we could store those repetitive calls in a dedicated impl at some point

Comment (antiyro) [reply crates/client/db/tests/test_block.rs]: same for state_diff_zero etc

File: crates/primitives/block/src/lib.rs [hunk 3]
```diff
@@ -151,7 +156,7 @@ impl DeoxysBlockInfo {
     }
 }
 
-#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+#[derive(Clone, Debug, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
```
Comment (antiyro) [inline crates/primitives/block/src/lib.rs:166]: Could you add some rustdoc to your new structures.
In the future we would like to have a proper rustdoc deployed.

**Review by antiyro — APPROVED** [review]

File: crates/client/db/tests/common/mod.rs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+use dc_db::DatabaseService;
+use dp_block::chain_config::ChainConfig;
+use tempfile::TempDir;
+
+pub async fn temp_db() -> DatabaseService {
+    let temp_dir = TempDir::new().unwrap();
+    let chain_config = std::sync::Arc::new(ChainConfig::test_config());
+    DatabaseService::new(temp_dir.path(), None, false, chain_config).await.unwrap()
```
Comment (EvolveArt) [inline crates/client/db/tests/common/mod.rs:8]: that's one of the reason I don't like `Option` or boolean arguments, like here without context on the struct we cannot tell what that means, maybe add some rust doc to help

File: crates/client/db/tests/test_block.rs [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+    backend.store_block(block_zero(), state_diff_zero(), vec![]).unwrap();
+    backend.store_block(block_pending_one(), state_diff_pending_one(), vec![]).unwrap();
+    backend.clear_pending_block().unwrap();
+
+    assert!(backend.get_block(&BLOCK_ID_PENDING).unwrap().is_none());
+
+    backend.store_block(block_one(), state_diff_one(), vec![]).unwrap();
+
+    let block = block_pending_two();
```
Comment (EvolveArt) [inline crates/client/db/tests/test_block.rs]: that's a bit misleading given that it also refers to a pending block for block number one

File: crates/client/db/tests/test_block.rs [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+    backend.store_block(block_zero(), state_diff_zero(), vec![]).unwrap();
+
+    let latest_block = block_one();
+    backend.store_block(latest_block.clone(), state_diff_one(), vec![]).unwrap();
+
+    assert_eq!(backend.get_latest_block_n().unwrap().unwrap(), 1);
+}
+
+#[tokio::test]
+async fn test_latest_confirmed_block() {
```
Comment (EvolveArt) [inline crates/client/db/tests/test_block.rs:138]: the interesting thing to test here are the edge cases, we should probably fuzz-test these to make sure it works with "all" values, not for this PR but worth creating an issue if you agree

File: crates/client/sync/src/commitments/mod.rs [hunk 7]
```diff
@@ -105,6 +107,27 @@ where
     bonsai_storage.root_hash(IDENTIFIER).expect("Failed to get root hash")
 }
 
+pub fn compute_root_keyed<H>(entries: &[(Felt, Felt)]) -> Felt
+where
+    H: StarkHash + Send + Sync,
+{
```
Comment (EvolveArt) [inline crates/client/sync/src/commitments/mod.rs:113]: not related to db tests right ?

File: crates/primitives/receipt/src/lib.rs [hunk 8]
```diff
@@ -20,6 +20,36 @@ pub enum TransactionReceipt {
...
+    fn from(receipt: DeployTransactionReceipt) -> Self {
+        TransactionReceipt::Deploy(receipt)
+    }
+}
+
+impl From<DeployAccountTransactionReceipt> for TransactionReceipt {
+    fn from(receipt: DeployAccountTransactionReceipt) -> Self {
+        TransactionReceipt::DeployAccount(receipt)
+    }
+}
```
Comment (EvolveArt) [inline crates/primitives/receipt/src/lib.rs:51]: wdyt about implementing `AsRef` too?

**Review by EvolveArt — COMMENTED** [review]
smol review

File: crates/client/sync/src/commitments/mod.rs (hunk 7 above)
Comment (jbcaron) [reply crates/client/sync/src/commitments/mod.rs:113]: I had added this for the *blockgen* crate (pseudo random block generation with a certain consistency), I removed this crate for the moment

File: crates/client/db/tests/test_block.rs (hunk 2 above)
Comment (jbcaron) [reply crates/client/db/tests/test_block.rs]: Yes, we will need a common logic with #215 : [test_utils.rs](https://github.com/madara-alliance/madara/blob/0140ed03e5812e2f3d807032f4e7fe76af861d3b/crates/client/rpc/src/test_utils.rs)

File: crates/primitives/receipt/src/lib.rs (hunk 8 above)
Comment (jbcaron) [reply crates/primitives/receipt/src/lib.rs:51]: do you think a `TryAsRef` on the enum or `AsRef` on the methods below?

Comment (EvolveArt) [reply crates/primitives/receipt/src/lib.rs:51]: hm `TryAsRef` should work there

Comment (jbcaron) [reply crates/primitives/receipt/src/lib.rs:51]: ok I will add that in #227 since it concerns primitives

File: crates/client/db/tests/test_block.rs (hunk 6 above)
Comment (jbcaron) [reply crates/client/db/tests/test_block.rs:138]: I don’t know if this is very relevant for DB tests, but for other crates, absolutely

File: crates/primitives/block/src/lib.rs (hunk 3 above)
Comment (jbcaron) [reply crates/primitives/block/src/lib.rs:166]: Done

File: crates/client/db/tests/common/mod.rs (hunk 4 above)
Comment (jbcaron) [reply crates/client/db/tests/common/mod.rs:8]: add doc, done.

**Review by EvolveArt — APPROVED** [review]
